okhttp/src/main/java/okhttp3/internal/connection/StreamAllocation.java

```diff
@@ -170,32 +172,44 @@ private RealConnection findConnection(int connectTimeout, int readTimeout, int w
         refusedStreamCount = 0;
       }
     }
-    RealConnection newConnection = new RealConnection(connectionPool, selectedRoute);
+
+    RealConnection result = new RealConnection(connectionPool, selectedRoute);
 
     synchronized (connectionPool) {
-      acquire(newConnection);
-      Internal.instance.put(connectionPool, newConnection);
-      this.connection = newConnection;
+      acquire(result);
+      Internal.instance.put(connectionPool, result);
```

Comment:
I'm wondering: should the put call occur after the deduplicate check? I think there is a possible (highly unlikely) set of interleavings that cause the `connections.allocations.size()` to be greater than 1, causing an ISE during the releaseAndAcquire.